clippy_lints/src/methods/search_is_some.rs

```diff
@@ -146,3 +152,122 @@ pub(super) fn check<'tcx>(
         }
     }
 }
+
+// Build suggestion gradually by handling closure arg specific usages,
+// such as explicit deref and borrowing cases.
+// Returns `None` if no such use cases have been triggered in closure body
+fn get_closure_suggestion<'tcx>(
+    cx: &LateContext<'_>,
+    search_arg: &'tcx hir::Expr<'_>,
+    closure_body: &hir::Body<'_>,
+) -> Option<String> {
+    let mut visitor = DerefDelegate {
+        cx,
+        closure_span: search_arg.span,
+        next_pos: None,
+        suggestion_start: String::new(),
+        suggestion_end: String::new(),
+    };
+
+    let fn_def_id = cx.tcx.hir().local_def_id(search_arg.hir_id);
+    cx.tcx.infer_ctxt().enter(|infcx| {
+        ExprUseVisitor::new(&mut visitor, &infcx, fn_def_id, cx.param_env, cx.typeck_results())
+            .consume_body(closure_body);
+    });
+
+    if visitor.suggestion_start.is_empty() {
+        None
+    } else {
+        Some(format!("{}{}", visitor.suggestion_start, visitor.suggestion_end))
+    }
+}
+
+struct DerefDelegate<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    closure_span: Span,
+    next_pos: Option<BytePos>,
+    suggestion_start: String,
+    suggestion_end: String,
+}
+
+impl<'tcx> Delegate<'tcx> for DerefDelegate<'_, 'tcx> {
+    fn consume(&mut self, _: &PlaceWithHirId<'tcx>, _: HirId) {}
+
+    fn borrow(&mut self, cmt: &PlaceWithHirId<'tcx>, _: HirId, _: ty::BorrowKind) {
+        if let PlaceBase::Local(id) = cmt.place.base {
+            let map = self.cx.tcx.hir();
+            let ident_str = map.name(id).to_string();
+            let span = map.span(cmt.hir_id);
+            let start_span = if let Some(next_pos) = self.next_pos {
+                Span::new(next_pos, span.lo(), span.ctxt())
+            } else {
+                self.closure_span.until(span)
+            };
+            let start_snip = snippet(self.cx, start_span, "..");
+            let end_span = Span::new(span.hi(), self.closure_span.hi(), span.ctxt());
+            let end_snip = snippet(self.cx, end_span, "..");
+
+            if cmt.place.projections.is_empty() {
+                // handle item without any projection, that needs an explicit borrowing
+                // i.e.: suggest `&x` instead of `x`
+                self.suggestion_start.push_str(&format!("{}&{}", start_snip, ident_str));
+            } else {
+                // cases where a parent call is using the item
+                // i.e.: suggest `.contains(&x)` for `.find(|x| [1, 2, 3].contains(x)).is_none()`
+                let parent_expr = get_parent_expr_for_hir(self.cx, cmt.hir_id);
+                if let Some(Expr { hir_id: _, kind, .. }) = parent_expr {
+                    if let ExprKind::Call(_, args) | ExprKind::MethodCall(_, _, args, _) = kind {
+                        let args_to_handle = args.iter().filter(|arg| arg.hir_id == cmt.hir_id).collect::<Vec<_>>();
```

Comment:
Ahh, okay, now I've also figured out what I was stuck on :upside_down_face:. I believe this logic can be simplified.

The `HirId` is unique for every expression/node. This means that the filter call on `args` in 235 will always only find one item. This can be even further simplified by not iterating over the arguments but using the linting context to get the expression, and it also removed the need for the `for`-loop and `if _.is_empty()` check. My refactoring looks like this (which should be equivalent):

```rs
if let Some(parent_expr) = get_parent_expr_for_hir(self.cx, cmt.hir_id) {
    if let ExprKind::Call(..) | ExprKind::MethodCall(..) = parent_expr.kind {
        let expr = self.cx.tcx.hir().expect_expr(cmt.hir_id);
        let arg_ty_kind = self.cx.typeck_results().expr_ty(expr).kind();

        // Note: this is should always be true, as `find` only gives us a reference which are not mutable
        if matches!(arg_ty_kind, ty::Ref(_, _, Mutability::Not)) {
            let start_span = Span::new(self.next_pos, span.lo(), span.ctxt());
            let start_snip = snippet_with_applicability(self.cx, start_span, "..", &mut self.applicability);

            self.suggestion_start.push_str(&format!("{}&{}", start_snip, ident_str));
            self.next_pos = span.hi();
        } else {
            self.applicability = Applicability::Unspecified;
        }
        return;
    }
}
```

Rust also automatically injects some derefs which we shouldn't suggest refs for. An example is the `self` argument, which gets dereferenced automatically (See [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c389d245a5d7849be83994673cf23609))

---

Reviewing this, I'm seeing how complex this task actually is. I kind of underestimated it. Thank you for putting in the effort :upside_down_face:

Reply:
Neat ! 😉  and thank you for the explanation